File: 3/3.php
<?php

class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        $count = strlen($s);
        $max = $len = 0;
        $arr = [];
        $i = $j = 0;
        while ($j < $count) {
            if (isset($arr[$s[$j]]) && $arr[$s[$j]] >= $i) {
                $i = $arr[$s[$j]];
                $len = $j - $i;
            } else {
                $len++;
            }
            if ($max < $len) $max = $len;
            $arr[$s[$j]] = $j;
            $j++;
        }
        return $max;
    }
}

var_dump((new Solution)->lengthOfLongestSubstring("abcabcbb"));

File: 3/3.go
package main

func main() {
	print(lengthOfLongestSubstring("au"))
}

func lengthOfLongestSubstring(s string) int {
	if len(s) < 2 {
		return len(s)
	}
	max, len, i := 0, 0, 0
	tmpMap := make(map[int32]int)
	for j, v := range s {
		if value, isset := tmpMap[v]; isset && value >= i {
			i = value
		}
		len = j - i + 1
		if max < len {
			max = len
		}
		tmpMap[v] = j + 1
	}
	return max
}
